### PR TITLE
Determines type of subject terms from FAST and include in cocina.

### LINKED ISSUE #1874: Determine type of subject terms from FAST and include in cocina
(cc @amyehodge )

Each FAST subject term falls into one of these categories:
* Personal Names (cocina type: person)
* Corporate Names (cocina type: organization)
* Event (cocina type: event)
* Uniform Titles (cocina type: title)
* Chronological (cocina type: time)
* Topical (cocina type: topic)
* Geographic (cocina type: place)
* FormGenre (cocina type: genre)
* Meeting (cocina type: conference)

It is important to know the category a given term belongs to because terms of different categories are treated in different ways in systems such as SearchWorks (such as the facet a subject term goes into, and what heading it displays under in the record). Currently all subject terms are being typed as "topic", which means that terms such as places, names, and titles will not be handled correctly.

H2 should determine the type of the subject in FAST and include the corresponding cocina subject.type according to the mapping above.

MODS mappings are already in place for the different subject types. Data created prior to resolving this issue will need to be reviewed and remediated.

### PR DESCRIPTION
**Pull request by justinlittman**
closes #1874

## Why was this change made?
Accurate subject metadata.

## How was this change tested?
Unit

## Which documentation and/or configurations were updated?
NA

### CODE REVIEW

File: app/controllers/autocomplete_controller.rb [hunk 1]
```diff
@@ -103,7 +104,34 @@ def sort_suggestions(suggestions)
...
 
-  def authority_uri(idroot)
-    "#{URI_PREFIX}#{idroot.delete_prefix('fst').to_i}/"
+  def key(idroot, type)
+    "#{URI_PREFIX}#{idroot.delete_prefix('fst').to_i}/::#{cocina_type(type)}"
   end
+
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/MethodLength
+  def cocina_type(type)
```
Comment (jcoyne) [inline app/controllers/autocomplete_controller.rb:113]: Can this be a Hash rather than a case statement?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/controllers/autocomplete_controller.rb (hunk 1 above)
Comment (justinlittman) [reply app/controllers/autocomplete_controller.rb:123]: Yup. Fixed.

File: app/components/works/keywords_row_component.html.erb [hunk 2]
```diff
@@ -3,7 +3,10 @@
     <%= form.label :label, 'Keyword', class: 'col-form-label sr-only' %>
     <div data-controller="autocomplete autocomplete-edit" data-autocomplete-url-value="/autocomplete" data-autocomplete-min-length-value="3" class="dropdown">
       <%= form.text_field :label, { class: "form-control#{ 'is-invalid' if error?}", required: true, 'data-autocomplete-target': 'input', 'data-autocomplete-edit-target': 'input' } %>
-      <%= form.hidden_field :uri, { required: true, 'data-autocomplete-target': 'hidden', 'data-autocomplete-edit-target': 'uri', 'data-action': 'autocomplete-edit#change' }%>
+      <%#= form.hidden_field :value, { 'data-autocomplete-target': 'hidden', 'data-autocomplete-edit-target': 'value', 'data-action': 'autocomplete-edit#change' }%>
```
Comment (jcoyne) [inline app/components/works/keywords_row_component.html.erb:6]: Is this cruft?

File: app/components/works/keywords_row_component.html.erb [hunk 3]
```diff
@@ -3,7 +3,10 @@
...
+      <input type="hidden" data-autocomplete-target="hidden" data-autocomplete-edit-target="value" data-action="autocomplete-edit#change" />
```
Comment (jcoyne) [inline app/components/works/keywords_row_component.html.erb:7]: What is the purpose of this? How can a hidden field trigger a change?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/packs/controllers/autocomplete_edit_controller.js [hunk 4]
```diff
@@ -1,15 +1,19 @@
...
-  static targets = ["uri", "input"]
+  static targets = ["uri", "input", "value", "type"]
 
   connect() {
     this.change()
   }
 
   change() {
-    if(this.uriTarget.value) {
+    console.log(this.valueTarget.value)
```
Comment (jcoyne) [inline app/packs/controllers/autocomplete_edit_controller.js:11]: remove this

Comment (justinlittman) [reply app/packs/controllers/autocomplete_edit_controller.js:11]: Duh.

File: app/components/works/keywords_row_component.html.erb (hunk 3 above)
Comment (justinlittman) [reply app/components/works/keywords_row_component.html.erb:6]: This hidden input is populated by the autocomplete controller (recall, an instance of stimulus-autocomplete). In turn this triggers the autocomplete-edit controller to split the value and populate the uri and cocina type inputs.

The autocomplete controller changing the hidden field does trigger the change -- not sure how.

Comment (jcoyne) [reply app/components/works/keywords_row_component.html.erb:6]: Can you add a comment about this mechanism?

Comment (justinlittman) [reply app/components/works/keywords_row_component.html.erb:8]: Done.

**Review by jcoyne — APPROVED** [review]

**Pull request merged**